skimage/segmentation/_join.py

```diff
@@ -56,7 +56,7 @@ def join_segmentations(s1, s2, return_mapping: bool = False):
     s1_relabeled, _, backward_map1 = relabel_sequential(s1)
     s2_relabeled, _, backward_map2 = relabel_sequential(s2)
     # Create joined label image
-    factor = s2.max() + 1
+    factor = (s2.max() + 1).astype(s2.dtype)
```

Comment:
Thanks, that is the trick! And it is working, but as you say, not for arrays.

I think the bigger risk for overflows is in the line of j_initial. While the limit of uint64 is hard to reach, it would more likely happen when the input has smaller dtypes. Then we would still could switch to alternative dtypes to prevent the overflow, if we cannot make it error.

```pycon
>>> # Example data
>>> dtype = np.uint8
>>> max_value = np.iinfo(dtype).max
255
>>> value = int(0.8 * max_value)  # Value triggering overflow only in j_initial
>>> value
205
>>> s1_relabeled = np.array([[0, value]], dtype)
>>> s2_relabeled = np.array([[0, value]], dtype)
>>> factor = value + s2_relabeled.dtype.type(1)
>>> # Workaround
>>> j_max = factor * s1_relabeled.max() + s2_relabeled.max()
>>> j_dtype = np.min_scalar_type(j_max)
>>> j_max, j_dtype
(42024, dtype('uint16'))
>>> # Failing test with current code
>>> j_initial = factor * s1_relabeled + s2_relabeled
>>> j_initial.max()
40
>>> assert j_initial.max() == j_max
AssertionError
>>> # With improved code
>>> j_initial = factor.astype(j_dtype) * s1_relabeled.astype(j_dtype) + s2_relabeled.astype(j_dtype)
>>> assert j_initial.max() == j_max

```
But I think scikit-image guarantees that input dtypes are preserved. Is this the case for this function as well?

Reply:
I should probably now for certain, but I think if we manipulate an array and return something that's directly derived from it we usually try to preserve the dtype. My instinct is to preserve dtype for this function as well if we can. That said, [`relabel_sequential`](https://github.com/scikit-image/scikit-image/blob/c459c2adfc2b15e616c6858e034de178f4ba449a/skimage/segmentation/_join.py#L162-L170) also seems to upcast if necessary. So let's use the same constraints for this function.

> j_max = factor * s1_relabeled.max() + s2_relabeled.max()

The calls to `max()` line might lead to quite a performance hit and it should be unnecessary as we can reason that
```python
assert s1_relabeled.max() == len(backward_map1) - 1
assert s2_relabeled.max() == len(backward_map2) - 1
```
That way we should be able to check for an overflow...